buffer/src/main/java/io/netty/buffer/AdaptivePoolingAllocator.java

```diff
@@ -605,52 +615,70 @@ private boolean allocate(int size, int sizeBucket, int maxCapacity, AdaptiveByte
+                    curr.readInitInto(buf, size, maxCapacity);
+                    current = curr;
+                    return true;
+                }
+
+                if (curr.remainingCapacity() == size) {
+                    // At this point we know that this will be the last time curr will be used, so directly set it to
+                    // null and release it once we are done.
+                    try {
+                        curr.readInitInto(buf, size, maxCapacity);
+                        return true;
+                    } finally {
+                        // Release in a finally block so even if readInitInto(...) would throw we would still correctly
+                        // release the current chunk before null it out.
+                        curr.release();
+                    }
                 } else {
-                    curr.attachToMagazine(this);
+                    // Release it as it's too small.
+                    curr.release();
                 }
             }
-            current = curr;
-            assert current != null;
-            if (last != null) {
-                if (last.remainingCapacity() < RETIRE_CAPACITY) {
-                    last.release();
+
+            // Now try to poll from the central queue first
+            curr = parent.centralQueue.poll();
+            if (curr == null) {
+                curr = newChunkAllocation(size);
+            } else {
+                if (curr.remainingCapacity() >= size) {
+                    // We have enough space so attach it.
+                    curr.attachToMagazine(this);
                 } else {
-                    transferChunk(last);
+                    // Check if we either retain the chunk in the nextInLine cache or releasing it.
+                    if (curr.remainingCapacity() < RETIRE_CAPACITY) {
+                        curr.release();
+                    } else {
+                        // See if it makes sense to transfer the Chunk to the nextInLine cache for later usage.
+                        // This method will release curr if this is not the case
+                        transferToNextInLineOrRelease(curr);
+                    }
+                    curr = newChunkAllocation(size);
                 }
             }
-            if (curr.remainingCapacity() > size) {
-                curr.readInitInto(buf, size, maxCapacity);
-            } else if (curr.remainingCapacity() == size) {
-                try {
+
+            current = curr;
+            try {
+                assert current.remainingCapacity() >= size;
+                if (curr.remainingCapacity() > size) {
                     curr.readInitInto(buf, size, maxCapacity);
-                } finally {
+                    this.current = curr;
+                    curr = null;
+                } else {
+                    curr.readInitInto(buf, size, maxCapacity);
+                    current = null;
```

Comment:
`current` will be assigned `null` in the finally-clause when we go through this branch